docs/core/diagnostics/diagnosticsource-diagnosticlistener.md

```diff
@@ -0,0 +1,343 @@
+
+Combining everything we have:
+
+```C#
+    class HttpClient
+    {
+        private static DiagnosticSource httpLogger= new DiagnosticListener("System.Net.Http");
+        public byte[] SendWebRequest(string url)
+        {
+            if (httpLogger.IsEnabled("RequestStart"))
+            {
+                httpLogger.Write("RequestStart", new { Url = url });
+            }
+            // pretend this sends an HTTP request to the url and gets back a reply
+            byte[] reply = new byte[] {};
+            return reply;
+        }
+    }
+```
+
+-------------------------------------------
+
+### Discovery of DiagnosticListeners
+
+The first step in receiving events is to discover which `DiagnosticListeners` you are
+interested in. `DiagnosticListener` supports a way of discovering `DiagnosticListeners` that are
+active in the system at runtime. The API to accomplish this is the `AllListeners`
+`IObservable<DiagnosticListener>`.
+
+Implement an `Observer<T>` class that inherits from the `IObservable` interface, which is the 'callback' version of the `IEnumerable` interface. You can learn more about it at the [Reactive Extensions](https://docs.microsoft.com/dotnet/csharp/fundamentals/types/anonymous-types) site.
+An `IObserver` has three callbacks, `OnNext`, `OnComplete`
+and `OnError`, and an `IObservable` has single method called `Subscribe` which gets passed one of these
+Observers. Once connected, the Observer gets callbacks (mostly `OnNext` callbacks) when things
+happen.
+
+A typical use of the `AllListeners` static property looks like this:
+
+```C#
+    class Observer<T> : IObserver<T>
+    {
+        public Observer(Action<T> onNext, Action onCompleted)
+        {
+            _onNext = onNext ?? new Action<T>(_ => { });
+            _onCompleted = onCompleted ?? new Action(() => { });
+        }
+
+        public void OnCompleted() { _onCompleted(); }
+        public void OnError(Exception error) { }
+        public void OnNext(T value) { _onNext(value); }
+
+        private Action<T> _onNext;
+        private Action _onCompleted;
+    }
+    
+    Action<DiagnosticListener> onNewListener = delegate (DiagnosticListener listener)
+    {
+        Console.WriteLine($"New Listener discovered: {listener.Name}");
+        if (listener.Name == "System.Net.Http")
+        {
+            // Here is where we put code to subscribe to the Listener.
+        }
+    };
+    IObserver<DiagnosticListener> observer = new Observer<DiagnosticListener>(onNewListener, null);
+    //when a listener is created, invoke the Observer onNext function which calls the delegate
+    listenerSubscription = DiagnosticListener.AllListeners.Subscribe(observer);
+
+    // Typically you leave the listenerSubscription subscription active forever.
+    // However when you no longer want your callback to be called, you can
+    // call listenerSubscription.Dispose() to cancel your subscription to the IObservable.
+```
```

Comment:
I think it would be better for teaching purposes to have a single code snippet with all the features implemented and commented. You can call out specific parts as you go, but having the implementation spread out as it is makes it difficult to follow and, more importantly, makes it difficult for users to copy/paste as a starting template for their own implementation.